main.go

```diff
@@ -0,0 +1,451 @@
+// Copyright 2018 Google Inc. All rights reserved.
+// Use of this source code is governed by the Apache 2.0
+// license that can be found in the LICENSE file.
+
+package main
+
+import (
+	"fmt"
+	"net/http"
+    "strconv"
+	"google.golang.org/appengine"
+    "google.golang.org/appengine/datastore"
+    "google.golang.org/appengine/log"
+)
+
+type Product struct {
+    Name string
+    Barcode string
+    Description string
+    Image string
+}
+
+type Price struct {
+    Barcode string
+    Price float64
+}
+
+type Stock struct {
+    Barcode string
+    Stock int64
+}
+
+type User struct {
+    Name string
+    Balance float64
+    Cart []string
+}
+
+func main() {
+	//http.HandleFunc("/", handle)
+    http.HandleFunc("/test/", testHandle)
+    http.HandleFunc("/products", productsHandle)
+    http.HandleFunc("/products/", productsHandle)
+    http.HandleFunc("/prices", pricesHandle)
+    http.HandleFunc("/prices/", pricesHandle)
+    http.HandleFunc("/stocks", stocksHandle)
+    http.HandleFunc("/stocks/", stocksHandle)
+    http.HandleFunc("/users", usersHandle)
+    http.HandleFunc("/users/", usersHandle)
+	appengine.Main()
+}
+
+/*
+func handle(w http.ResponseWriter, r *http.Request) {
+
+    fmt.Fprintln(w, "home")
+
+}
+*/
+
+func testHandle(w http.ResponseWriter, r *http.Request) {
+    fmt.Fprintln(w, r.URL.Path)
+
+}
+
+func productsHandle(w http.ResponseWriter, r *http.Request) {
+    ctx := appengine.NewContext(r)
+    switch r.Method{
+    case "GET":
+        q := datastore.NewQuery("Product")
+        var products []Product
+        keys, err := q.GetAll(ctx, &products)
+        if err != nil {
+            log.Errorf(ctx, "fetching products: %v", err)
+            return
+        }
+        if len(r.URL.Path) > len("/products/") {
+            targetCode := r.URL.Path[len("/products/"):]
+            var targetProduct Product
+            for i, _ := range keys {
```

Comment:
This code is bad.   If you have 1 million product,  even only query a specific product,  you are reading 1 million first, and check one by one.   

Let datastore do the production selection.  I think it has a way for you to specify which product to retrieve.